File: src/ugv_teleop/src/keyboard_teleop_core.cpp
#include "ugv_teleop/keyboard_teleop_core.hpp"

#include <algorithm>
#include <array>
#include <cctype>
#include <cmath>

namespace ugv_teleop
{

KeyboardTeleopCore::KeyboardTeleopCore(
  double linear_speed,
  double angular_speed,
  double speed_step,
  double turn_step,
  double accel_limit_linear,
  double decel_limit_linear,
  double accel_limit_angular,
  double decel_limit_angular,
  double idle_timeout_sec)
: linear_speed_(std::max(0.0, linear_speed)),
  angular_speed_(std::max(0.0, angular_speed)),
  speed_step_(std::max(0.0, speed_step)),
  turn_step_(std::max(0.0, turn_step)),
  accel_limit_linear_(std::max(1e-6, accel_limit_linear)),
  decel_limit_linear_(std::max(1e-6, decel_limit_linear)),
  accel_limit_angular_(std::max(1e-6, accel_limit_angular)),
  decel_limit_angular_(std::max(1e-6, decel_limit_angular)),
  idle_timeout_sec_(std::max(0.0, idle_timeout_sec)),
  last_input_time_(std::chrono::steady_clock::now())
{
}

bool KeyboardTeleopCore::handle_key_press(const std::string & key, std::optional<TimePoint> now)
{
  const auto normalized = normalize_key(key);
  if (!normalized.has_value()) {
    return false;
  }

  const auto timestamp = coerce_now(now);
  std::lock_guard<std::mutex> lock(mutex_);
  last_input_time_ = timestamp;

  if (normalized.value() == "x") {
    clear_move_keys_locked();
    zero_command_locked();
    last_update_time_ = timestamp;
    return true;
  }

  const auto & move_binding_map = move_bindings();
  if (move_binding_map.find(normalized.value()) != move_binding_map.end()) {
    pressed_move_keys_.insert(normalized.value());
    move_key_last_seen_[normalized.value()] = timestamp;
    return true;
  }

  const auto & speed_binding_map = speed_bindings();
  const auto speed_it = speed_binding_map.find(normalized.value());
  if (speed_it != speed_binding_map.end()) {
    linear_speed_ = std::max(
      0.0, linear_speed_ + speed_it->second.linear_step_direction * speed_step_);
    angular_speed_ = std::max(
      0.0, angular_speed_ + speed_it->second.angular_step_direction * turn_step_);
    return true;
  }

  return false;
}

bool KeyboardTeleopCore::handle_key_release(const std::string & key, std::optional<TimePoint> now)
{
  const auto normalized = normalize_key(key);
  if (!normalized.has_value()) {
    return false;
  }

  const auto & move_binding_map = move_bindings();
  if (move_binding_map.find(normalized.value()) == move_binding_map.end()) {
    return false;
  }

  const auto timestamp = coerce_now(now);
  std::lock_guard<std::mutex> lock(mutex_);
  last_input_time_ = timestamp;
  if (normalized.value() != "x") {
    pressed_move_keys_.erase(normalized.value());
    move_key_last_seen_.erase(normalized.value());
  }
  return true;
}

void KeyboardTeleopCore::clear_move_keys(std::optional<TimePoint> now)
{
  const auto timestamp = coerce_now(now);
  std::lock_guard<std::mutex> lock(mutex_);
  clear_move_keys_locked();
  last_input_time_ = timestamp;
}

void KeyboardTeleopCore::emergency_stop(std::optional<TimePoint> now)
{
  const auto timestamp = coerce_now(now);
  std::lock_guard<std::mutex> lock(mutex_);
  clear_move_keys_locked();
  zero_command_locked();
  last_input_time_ = timestamp;
  last_update_time_ = timestamp;
}

void KeyboardTeleopCore::expire_stale_move_keys(
  double stale_after_sec,
  std::optional<TimePoint> now)
{
  const auto timeout = std::max(0.0, stale_after_sec);
  const auto timestamp = coerce_now(now);
  std::lock_guard<std::mutex> lock(mutex_);
  expire_stale_locked(timestamp, timeout);
}

TeleopCommand KeyboardTeleopCore::snapshot(
  std::optional<TimePoint> now,
  std::optional<double> stale_key_timeout_sec)
{
  const auto timestamp = coerce_now(now);
  std::lock_guard<std::mutex> lock(mutex_);

  if (stale_key_timeout_sec.has_value()) {
    expire_stale_locked(timestamp, std::max(0.0, stale_key_timeout_sec.value()));
  }

  const auto idle_elapsed =
    std::chrono::duration<double>(timestamp - last_input_time_).count();
  if (idle_timeout_sec_ > 0.0 && idle_elapsed > idle_timeout_sec_) {
    clear_move_keys_locked();
  }

  const auto target = target_vector_locked();
  const double dt = last_update_time_.has_value() ?
    std::max(
    0.0,
    std::chrono::duration<double>(timestamp - last_update_time_.value()).count()) : 0.0;

  const double target_linear_x = target.x * linear_speed_;
  const double target_linear_y = target.y * linear_speed_;
  const double target_linear_z = target.z * linear_speed_;
  const double target_angular_z = target.th * angular_speed_;

  current_linear_x_ = slew_axis(
    current_linear_x_,
    target_linear_x,
    dt,
    accel_limit_linear_,
    decel_limit_linear_);
  current_linear_y_ = slew_axis(
    current_linear_y_,
    target_linear_y,
    dt,
    accel_limit_linear_,
    decel_limit_linear_);
  current_linear_z_ = slew_axis(
    current_linear_z_,
    target_linear_z,
    dt,
    accel_limit_linear_,
    decel_limit_linear_);
  current_angular_z_ = slew_axis(
    current_angular_z_,
    target_angular_z,
    dt,
    accel_limit_angular_,
    decel_limit_angular_);
  last_update_time_ = timestamp;

  return TeleopCommand{
    current_linear_x_,
    current_linear_y_,
    current_linear_z_,
    current_angular_z_,
    linear_speed_,
    angular_speed_,
    sorted_keys(pressed_move_keys_)};
}

KeyboardTeleopStatus KeyboardTeleopCore::status() const
{
  std::lock_guard<std::mutex> lock(mutex_);
  return KeyboardTeleopStatus{
    linear_speed_,
    angular_speed_,
    sorted_keys(pressed_move_keys_)};
}

TeleopCommand KeyboardTeleopCore::zero_command()
{
  std::lock_guard<std::mutex> lock(mutex_);
  zero_command_locked();
  return TeleopCommand{
    0.0,
    0.0,
    0.0,
    0.0,
    linear_speed_,
    angular_speed_,
    sorted_keys(pressed_move_keys_)};
}

bool KeyboardTeleopCore::is_move_key(const std::string & key)
{
  const auto normalized = normalize_key(key);
  if (!normalized.has_value()) {
    return false;
  }
  return move_bindings().find(normalized.value()) != move_bindings().end();
}

bool KeyboardTeleopCore::is_speed_key(const std::string & key)
{
  const auto normalized = normalize_key(key);
  if (!normalized.has_value()) {
    return false;
  }
  return speed_bindings().find(normalized.value()) != speed_bindings().end();
}

std::optional<std::string> KeyboardTeleopCore::normalize_key(const std::string & key)
{
  if (key.empty()) {
    return std::nullopt;
  }
  std::string normalized = key;
  std::transform(
    normalized.begin(),
    normalized.end(),
    normalized.begin(),
    [](unsigned char c) {return static_cast<char>(std::tolower(c));});
  if (
    move_bindings().find(normalized) != move_bindings().end() ||
    speed_bindings().find(normalized) != speed_bindings().end())
  {
    return normalized;
  }
  return std::nullopt;
}

double KeyboardTeleopCore::slew_axis(
  double current,
  double target,
  double dt,
  double accel_limit,
  double decel_limit)
{
  if (dt <= 0.0) {
    return target;
  }
  if (std::fabs(current - target) <= 1e-9) {
    return target;
  }

  double limit = accel_limit;
  if (current == 0.0) {
    limit = std::fabs(target) > 0.0 ? accel_limit : decel_limit;
  } else if ((current * target) < 0.0 || std::fabs(target) < std::fabs(current)) {
    limit = decel_limit;
  }

  const double max_delta = limit * dt;
  const double delta = target - current;
  if (std::fabs(delta) <= max_delta) {
    return target;
  }
  return current + std::copysign(max_delta, delta);
}

KeyboardTeleopCore::TimePoint KeyboardTeleopCore::coerce_now(std::optional<TimePoint> now)
{
  return now.has_value() ? now.value() : std::chrono::steady_clock::now();
}

const std::map<std::string, KeyboardTeleopCore::MoveVector> & KeyboardTeleopCore::move_bindings()
{
  static const std::map<std::string, MoveVector> kBindings = {
    {"w", MoveVector{1.0, 0.0, 0.0, 0.0}},
    {"s", MoveVector{-1.0, 0.0, 0.0, 0.0}},
    {"a", MoveVector{0.0, 1.0, 0.0, 0.0}},
    {"d", MoveVector{0.0, -1.0, 0.0, 0.0}},
    {"q", MoveVector{0.0, 0.0, 0.0, 1.0}},
    {"e", MoveVector{0.0, 0.0, 0.0, -1.0}},
    {"x", MoveVector{0.0, 0.0, 0.0, 0.0}},
  };
  return kBindings;
}

const std::map<std::string, KeyboardTeleopCore::SpeedVector> & KeyboardTeleopCore::speed_bindings()
{
  static const std::map<std::string, SpeedVector> kBindings = {
    {"i", SpeedVector{1.0, 0.0}},
    {"k", SpeedVector{-1.0, 0.0}},
    {"o", SpeedVector{0.0, 1.0}},
    {"l", SpeedVector{0.0, -1.0}},
  };
  return kBindings;
}

void KeyboardTeleopCore::clear_move_keys_locked()
{
  pressed_move_keys_.clear();
  move_key_last_seen_.clear();
}

void KeyboardTeleopCore::zero_command_locked()
{
  current_linear_x_ = 0.0;
  current_linear_y_ = 0.0;
  current_linear_z_ = 0.0;
  current_angular_z_ = 0.0;
}

void KeyboardTeleopCore::expire_stale_locked(TimePoint now, double timeout_sec)
{
  std::vector<std::string> stale_keys;
  stale_keys.reserve(move_key_last_seen_.size());
  for (const auto & [key, seen_at] : move_key_last_seen_) {
    const double elapsed = std::chrono::duration<double>(now - seen_at).count();
    if (elapsed > timeout_sec) {
      stale_keys.push_back(key);
    }
  }
  for (const auto & key : stale_keys) {
    pressed_move_keys_.erase(key);
    move_key_last_seen_.erase(key);
  }
}

KeyboardTeleopCore::MoveVector KeyboardTeleopCore::target_vector_locked() const
{
  MoveVector total;
  for (const auto & key : pressed_move_keys_) {
    const auto it = move_bindings().find(key);
    if (it == move_bindings().end()) {
      continue;
    }
    total.x += it->second.x;
    total.y += it->second.y;
    total.z += it->second.z;
    total.th += it->second.th;
  }

  total.x = std::clamp(total.x, -1.0, 1.0);
  total.y = std::clamp(total.y, -1.0, 1.0);
  total.z = std::clamp(total.z, -1.0, 1.0);
  total.th = std::clamp(total.th, -1.0, 1.0);
  return total;
}

std::vector<std::string> KeyboardTeleopCore::sorted_keys(const std::set<std::string> & keys)
{
  return std::vector<std::string>(keys.begin(), keys.end());
}

}

File: src/ugv_teleop/ugv_teleop/keyboard_teleop_core.py
import math
import threading
import time
from dataclasses import dataclass


MOVE_BINDINGS = {
    "w": (1.0, 0.0, 0.0, 0.0),
    "s": (-1.0, 0.0, 0.0, 0.0),
    "a": (0.0, 1.0, 0.0, 0.0),
    "d": (0.0, -1.0, 0.0, 0.0),
    "q": (0.0, 0.0, 0.0, 1.0),
    "e": (0.0, 0.0, 0.0, -1.0),
    "x": (0.0, 0.0, 0.0, 0.0),
}

SPEED_BINDINGS = {
    "i": (1.0, 0.0),
    "k": (-1.0, 0.0),
    "o": (0.0, 1.0),
    "l": (0.0, -1.0),
}


@dataclass(frozen=True)
class TeleopCommand:
    linear_x: float
    linear_y: float
    linear_z: float
    angular_z: float
    linear_speed: float
    angular_speed: float
    active_keys: tuple[str, ...]


class KeyboardTeleopCore:
    def __init__(
        self,
        *,
        linear_speed: float,
        angular_speed: float,
        speed_step: float,
        turn_step: float,
        accel_limit_linear: float,
        decel_limit_linear: float,
        accel_limit_angular: float,
        decel_limit_angular: float,
        idle_timeout_sec: float,
    ) -> None:
        self._lock = threading.Lock()
        self._pressed_move_keys: set[str] = set()
        self._move_key_last_seen: dict[str, float] = {}
        self._linear_speed = max(0.0, float(linear_speed))
        self._angular_speed = max(0.0, float(angular_speed))
        self._speed_step = max(0.0, float(speed_step))
        self._turn_step = max(0.0, float(turn_step))
        self._accel_limit_linear = max(1e-6, float(accel_limit_linear))
        self._decel_limit_linear = max(1e-6, float(decel_limit_linear))
        self._accel_limit_angular = max(1e-6, float(accel_limit_angular))
        self._decel_limit_angular = max(1e-6, float(decel_limit_angular))
        self._idle_timeout_sec = max(0.0, float(idle_timeout_sec))

        self._current_linear_x = 0.0
        self._current_linear_y = 0.0
        self._current_linear_z = 0.0
        self._current_angular_z = 0.0
        self._last_input_time = time.monotonic()
        self._last_update_time = None

    def handle_key_press(self, key: str, now: float | None = None) -> bool:
        key = self._normalize_key(key)
        if key is None:
            return False

        timestamp = self._coerce_now(now)
        with self._lock:
            self._last_input_time = timestamp
            if key == "x":
                self._clear_move_keys_locked()
                self._zero_command_locked()
                self._last_update_time = timestamp
                return True

            if key in MOVE_BINDINGS:
                self._pressed_move_keys.add(key)
                self._move_key_last_seen[key] = timestamp
                return True

            if key in SPEED_BINDINGS:
                linear_step_direction, angular_step_direction = SPEED_BINDINGS[key]
                self._linear_speed = max(0.0, self._linear_speed + linear_step_direction * self._speed_step)
                self._angular_speed = max(0.0, self._angular_speed + angular_step_direction * self._turn_step)
                return True

        return False

    def handle_key_release(self, key: str, now: float | None = None) -> bool:
        key = self._normalize_key(key)
        if key is None or key not in MOVE_BINDINGS:
            return False

        timestamp = self._coerce_now(now)
        with self._lock:
            self._last_input_time = timestamp
            if key != "x":
                self._pressed_move_keys.discard(key)
                self._move_key_last_seen.pop(key, None)
        return True

    def clear_move_keys(self, now: float | None = None) -> None:
        timestamp = self._coerce_now(now)
        with self._lock:
            self._clear_move_keys_locked()
            self._last_input_time = timestamp

    def emergency_stop(self, now: float | None = None) -> None:
        timestamp = self._coerce_now(now)
        with self._lock:
            self._clear_move_keys_locked()
            self._zero_command_locked()
            self._last_input_time = timestamp
            self._last_update_time = timestamp

    def expire_stale_move_keys(self, stale_after_sec: float, now: float | None = None) -> None:
        timeout = max(0.0, float(stale_after_sec))
        timestamp = self._coerce_now(now)
        with self._lock:
            stale_keys = [
                key for key, seen_at in self._move_key_last_seen.items()
                if (timestamp - seen_at) > timeout
            ]
            for key in stale_keys:
                self._pressed_move_keys.discard(key)
                self._move_key_last_seen.pop(key, None)

    def snapshot(self, now: float | None = None, stale_key_timeout: float | None = None) -> TeleopCommand:
        timestamp = self._coerce_now(now)
        with self._lock:
            if stale_key_timeout is not None:
                self._expire_stale_locked(timestamp, max(0.0, float(stale_key_timeout)))

            if self._idle_timeout_sec > 0.0 and (timestamp - self._last_input_time) > self._idle_timeout_sec:
                self._clear_move_keys_locked()

            target_x, target_y, target_z, target_th = self._target_vector_locked()
            dt = 0.0 if self._last_update_time is None else max(0.0, timestamp - self._last_update_time)

            target_linear_x = target_x * self._linear_speed
            target_linear_y = target_y * self._linear_speed
            target_linear_z = target_z * self._linear_speed
            target_angular_z = target_th * self._angular_speed

            self._current_linear_x = self._slew_axis(
                self._current_linear_x,
                target_linear_x,
                dt,
                self._accel_limit_linear,
                self._decel_limit_linear,
            )
            self._current_linear_y = self._slew_axis(
                self._current_linear_y,
                target_linear_y,
                dt,
                self._accel_limit_linear,
                self._decel_limit_linear,
            )
            self._current_linear_z = self._slew_axis(
                self._current_linear_z,
                target_linear_z,
                dt,
                self._accel_limit_linear,
                self._decel_limit_linear,
            )
            self._current_angular_z = self._slew_axis(
                self._current_angular_z,
                target_angular_z,
                dt,
                self._accel_limit_angular,
                self._decel_limit_angular,
            )
            self._last_update_time = timestamp

            return TeleopCommand(
                linear_x=self._current_linear_x,
                linear_y=self._current_linear_y,
                linear_z=self._current_linear_z,
                angular_z=self._current_angular_z,
                linear_speed=self._linear_speed,
                angular_speed=self._angular_speed,
                active_keys=tuple(sorted(self._pressed_move_keys)),
            )

    def status(self) -> dict[str, object]:
        with self._lock:
            return {
                "linear_speed": self._linear_speed,
                "angular_speed": self._angular_speed,
                "active_keys": tuple(sorted(self._pressed_move_keys)),
            }

    def zero_command(self) -> TeleopCommand:
        with self._lock:
            self._zero_command_locked()
            active_keys = tuple(sorted(self._pressed_move_keys))
            return TeleopCommand(
                linear_x=0.0,
                linear_y=0.0,
                linear_z=0.0,
                angular_z=0.0,
                linear_speed=self._linear_speed,
                angular_speed=self._angular_speed,
                active_keys=active_keys,
            )

    def _target_vector_locked(self) -> tuple[float, float, float, float]:
        x = 0.0
        y = 0.0
        z = 0.0
        th = 0.0
        for key in self._pressed_move_keys:
            dx, dy, dz, dth = MOVE_BINDINGS[key]
            x += dx
            y += dy
            z += dz
            th += dth

        return (
            max(-1.0, min(1.0, x)),
            max(-1.0, min(1.0, y)),
            max(-1.0, min(1.0, z)),
            max(-1.0, min(1.0, th)),
        )

    def _clear_move_keys_locked(self) -> None:
        self._pressed_move_keys.clear()
        self._move_key_last_seen.clear()

    def _zero_command_locked(self) -> None:
        self._current_linear_x = 0.0
        self._current_linear_y = 0.0
        self._current_linear_z = 0.0
        self._current_angular_z = 0.0

    def _expire_stale_locked(self, now: float, timeout: float) -> None:
        stale_keys = [
            key for key, seen_at in self._move_key_last_seen.items()
            if (now - seen_at) > timeout
        ]
        for key in stale_keys:
            self._pressed_move_keys.discard(key)
            self._move_key_last_seen.pop(key, None)

    @staticmethod
    def _normalize_key(key: str | None) -> str | None:
        if not key:
            return None
        normalized = key.lower()
        if normalized in MOVE_BINDINGS or normalized in SPEED_BINDINGS:
            return normalized
        return None

    @staticmethod
    def _coerce_now(now: float | None) -> float:
        return time.monotonic() if now is None else float(now)

    @staticmethod
    def _slew_axis(current: float, target: float, dt: float, accel_limit: float, decel_limit: float) -> float:
        if dt <= 0.0:
            return target
        if math.isclose(current, target, abs_tol=1e-9):
            return target

        if current == 0.0:
            limit = accel_limit if abs(target) > 0.0 else decel_limit
        elif current * target < 0.0 or abs(target) < abs(current):
            limit = decel_limit
        else:
            limit = accel_limit

        max_delta = limit * dt
        delta = target - current
        if abs(delta) <= max_delta:
            return target
        return current + math.copysign(max_delta, delta)
